bundles/org.openhab.binding.mycroft/src/main/java/org/openhab/binding/mycroft/internal/api/MycroftConnection.java

```diff
@@ -0,0 +1,268 @@
+
+    private final WebSocketClient client;
+    private final String socketName;
+    private final Gson gson;
+
+    private final MycroftConnectionListener connectionListener;
+    private final Map<MessageType, Set<MycroftMessageListener<? extends BaseMessage>>> listeners = new ConcurrentHashMap<>();
+
+    private ConnectionState connectionState = ConnectionState.DISCONNECTED;
+    private @Nullable Session session;
+
+    private static final int TIMEOUT_MILLISECONDS = 3000;
+
+    public MycroftConnection(MycroftConnectionListener listener, WebSocketClient client) {
+        this.connectionListener = listener;
+        this.client = client;
+        this.client.setConnectTimeout(TIMEOUT_MILLISECONDS);
+        this.client.setMaxIdleTimeout(0);
+        this.socketName = "Websocket$" + System.currentTimeMillis() + "-" + INSTANCE_COUNTER.incrementAndGet();
+
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(MessageType.class, new MessageTypeConverter());
+        gson = gsonBuilder.create();
+    }
+
+    public MycroftConnection(MycroftConnectionListener listener) {
+        this.connectionListener = listener;
+        this.client = new WebSocketClient();
+        this.client.setMaxIdleTimeout(0);
+        this.client.setConnectTimeout(TIMEOUT_MILLISECONDS);
+        this.socketName = "Websocket$" + System.currentTimeMillis() + "-" + INSTANCE_COUNTER.incrementAndGet();
+
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(MessageType.class, new MessageTypeConverter());
+        gson = gsonBuilder.create();
+    }
+
+    public void start(String ip, int port) {
+        if (connectionState == ConnectionState.CONNECTED) {
+            return;
+        } else if (connectionState == ConnectionState.CONNECTING) {
+            logger.debug("{} already connecting", socketName);
+            return;
+        } else if (connectionState == ConnectionState.DISCONNECTING) {
+            logger.warn("{} trying to re-connect while still disconnecting", socketName);
+        }
+        Future<Session> futureConnect = null;
+        try {
+            URI destUri = URI.create("ws://" + ip + ":" + port + "/core");
+            client.start();
+            logger.debug("Trying to connect {} to {}", socketName, destUri);
+            futureConnect = client.connect(this, destUri);
+            futureConnect.get(TIMEOUT_MILLISECONDS, TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
+            if (futureConnect != null) {
+                futureConnect.cancel(true);
+            }
+            connectionListener
+                    .connectionLost("Error while connecting: " + (e.getMessage() != null ? e.getMessage() : "unknown"));
+        }
+    }
+
+    public void close() {
+        try {
+            connectionState = ConnectionState.DISCONNECTING;
+            client.stop();
+        } catch (Exception e) {
```

Comment:
see above

Reply:
Same here : jetty doesn't let me the choice.

Reply:
If that is the case, I guess catching `Exception` is fine as the library (in this case jetty) doesn't give you a choice.